src/storage/src/source/kafka.rs

```diff
@@ -441,17 +452,26 @@ impl SourceRender for KafkaSourceConnection {
                         }
                     }
 
-                    for &pid in partitions.keys() {
+                    for (&pid, &upper) in &partitions {
                         if config.responsible_for(pid) {
                             reader.ensure_partition(pid);
-                            let part_min_ts = Partitioned::with_partition(pid, MzOffset::from(0));
-                            reader
-                                .partition_capabilities
-                                .entry(pid)
-                                .or_insert_with(|| data_cap.delayed(&part_min_ts));
+                            if let Entry::Vacant(entry) = reader.partition_capabilities.entry(pid) {
+                                let start_offset = match reader.start_offsets.get(&pid) {
+                                    Some(&offset) => offset.try_into().unwrap(),
+                                    None => 0u64,
+                                };
+                                let part_min_ts = Partitioned::with_partition(pid, MzOffset::from(start_offset));
+                                let upper_offset = MzOffset::from(u64::try_from(upper).expect("invalid negative offset"));
+                                let part_upper_ts = Partitioned::with_partition(pid, upper_offset);
+                                entry.insert(PartitionCapability {
+                                    data: data_cap.delayed(&part_min_ts),
+                                    progress: progress_cap.delayed(&part_upper_ts),
+                                });
```

Comment:
this block is worth a comment; its the core logic of this pr, where we downgrade the `progress_cap` to the upper of the topic

Reply:
Done